src/main/java/ca/corefacility/bioinformatics/irida/service/remote/ProjectHashingService.java

```diff
@@ -0,0 +1,94 @@
+package ca.corefacility.bioinformatics.irida.service.remote;
+
+import ca.corefacility.bioinformatics.irida.model.joins.Join;
+import ca.corefacility.bioinformatics.irida.model.joins.impl.SampleGenomeAssemblyJoin;
+import ca.corefacility.bioinformatics.irida.model.project.Project;
+import ca.corefacility.bioinformatics.irida.model.sample.Sample;
+import ca.corefacility.bioinformatics.irida.model.sample.SampleSequencingObjectJoin;
+import ca.corefacility.bioinformatics.irida.model.sample.metadata.MetadataEntry;
+import ca.corefacility.bioinformatics.irida.service.GenomeAssemblyService;
+import ca.corefacility.bioinformatics.irida.service.ProjectService;
+import ca.corefacility.bioinformatics.irida.service.SequencingObjectService;
+import ca.corefacility.bioinformatics.irida.service.sample.SampleService;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Component;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Component used to compute a deep hashcode of a full project.  Looks at all project details, samples, metadata, and sequencing data within the project to generate a full hashcode.
+ */
+@Component
+public class ProjectHashingService {
+    private ProjectService projectService;
+    private SampleService sampleService;
+    private SequencingObjectService sequencingObjectService;
+    private GenomeAssemblyService assemblyService;
+
+    @Autowired
+    public ProjectHashingService(ProjectService projectService, SampleService sampleService, SequencingObjectService sequencingObjectService, GenomeAssemblyService assemblyService) {
+        this.projectService = projectService;
+        this.sampleService = sampleService;
+        this.sequencingObjectService = sequencingObjectService;
+        this.assemblyService = assemblyService;
+    }
+
+    /**
+     * Get a deep hashsum for the full project
+     *
+     * @param project the {@link Project} to generate a hashsum for
+     * @return the computed hashsum
+     */
+    @PreAuthorize("hasPermission(#project, 'canReadProject')")
+    public Integer getProjectHash(Project project) {
+
+        HashCodeBuilder builder = new HashCodeBuilder();
+
+        builder.append(project);
```

Comment:
A question about the builder.append. Should we maybe have these calls in a try/catch block? In case there any exceptions are thrown by the hashcodebuilder? That way at least we can log the exception rather than it being swallowed up.

Reply:
RIght now if an exception gets thrown while hashing it won't get swallowed, it'll go un-caught up to the top.  I think this is probably for the best since then it'll run up to the webserver which will give a 500 error and log the whole stack trace.  If there's any exceptions building the hash I think the right thing to do is to error out so I wouldn't want to just catch and log.

Other option would be to throw a new custom exception for this, but since there isn't really anything special to do to handle it I'm not sure it's worth it.  Thoughts?

Reply:
Oh ok well as long as the error is swallowed then it's all good! Yeah not worth it in that case. Thanks